### PR TITLE
When patching test class, don't run the original tests

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103523

The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: test/dynamo/test_dynamic_shapes.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def make_dynamic_cls(cls):
 ]
 for test in tests:
     make_dynamic_cls(test)
+del globals()["test"]
```
Comment (albanD) [inline test/dynamo/test_dynamic_shapes.py:71]: del test should be enough here no?

File: test/dynamo/test_higher_order_ops.py [hunk 2]
```diff
@@ -1614,7 +1614,7 @@ def run(self, result=None):
         # support all transforms, so manually patch it to `True`` for
         # testing on release branch.
         with config.patch(capture_func_transforms=True):
-            super().run(result)
+            super(type(self), self).run(result)
```
Comment (albanD) [inline test/dynamo/test_higher_order_ops.py:1617]: Why is this needed here? Isn't this never needed in 3.8+

File: torch/_dynamo/testing.py [hunk 3]
```diff
@@ -317,23 +317,24 @@ def _fn(*args, **kwargs):
...
+                setattr(DummyTestClass, name, getattr(cls, name))
                 continue
             new_name = f"{name}{fn_suffix}"
             new_fn = _make_fn_with_patches(fn, *patches)
             new_fn.__name__ = new_name
             if xfail_prop is not None and hasattr(fn, xfail_prop):
                 new_fn = unittest.expectedFailure(new_fn)
             setattr(DummyTestClass, new_name, new_fn)
+        # NB: Doesn't handle slots correctly, but whatever
+        elif name not in DummyTestClass.__dict__:
```
Comment (albanD) [inline torch/_dynamo/testing.py:336]:
```suggestion
        elif hasattr(DummyTestClass, name):
```

Would that solve the slot problem?

File: torch/_dynamo/testing.py [hunk 4]
```diff
@@ -317,23 +317,24 @@ def _fn(*args, **kwargs):
...
 def make_test_cls_with_patches(cls, cls_prefix, fn_suffix, *patches, xfail_prop=None):
-    class DummyTestClass(cls):
-        pass
-
-    DummyTestClass.__name__ = f"{cls_prefix}{cls.__name__}"
+    DummyTestClass = type(f"{cls_prefix}{cls.__name__}", cls.__bases__, {})
     DummyTestClass.__qualname__ = DummyTestClass.__name__
 
-    for name in dir(cls):
+    for name in cls.__dict__.keys():
```
Comment (albanD) [inline torch/_dynamo/testing.py:323]: Why not dir() anymore?

File: torch/_dynamo/testing.py (hunk 3 above)
Comment (clee2000) [reply torch/_dynamo/testing.py]: Is there a preference here for has vs not has? I assumed we would want to copy over the things that don't exist

**Comment by ezyang** [comment]
@clee2000 feel free to resubmit under your name so I can give you review juice
